jrt/src/com/yahoo/jrt/SslSocket.java

```diff
@@ -0,0 +1,232 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.jrt;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLEngineResult;
+import javax.net.ssl.SSLException;
+import javax.net.ssl.SSLSession;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SocketChannel;
+import java.util.logging.Logger;
+
+import static javax.net.ssl.SSLEngineResult.*;
+
+/**
+ * A {@link CryptoSocket} using TLS ({@link SSLEngine})
+ *
+ * @author bjorncs
+ */
+public class SslSocket implements CryptoSocket {
+
+    private static final Logger log = Logger.getLogger(SslSocket.class.getName());
+
+    private enum HandshakeState { NOT_STARTED, NEED_READ, NEED_WRITE, COMPLETED }
+
+    private final SocketChannel channel;
+    private final SSLEngine sslEngine;
+    private final ByteBuffer wrapBuffer;
```

Comment:
consider using jrt.Buffer class for buffers to not need to keep track of read/write mode and content size/location (for future consideration, not blocking this PR)

Reply:
I have changed `wrapBuffer` and `unwrapBuffer` to use `com.yahoo.jrt.Buffer` in latest commit